### PR TITLE
Adding support for nonzero in LazyTensor shape functions

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77572

### CODE REVIEW

**Comment by Gamrix** [comment]
cc @miladm

**Review by Krovatkin — COMMENTED** [review]
Could you please add a test case that would test this approach works. Namely,
1. we can set a flag
2. trace nonzero
3. it gives us the correct upper bound
4. when we move a lazy nonzero to `cpu()` it gives us the correct result.

File: torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp [hunk 1]
```diff
@@ -153,6 +153,12 @@ bool force_eager_fallback(c10::Symbol op) {
       return true;
     }
   }
+  if(op == at::aten::nonzero){
+    // When symbolic shape mode is not enabled, the nonzero shape function
+    // returns an incorrect result.
+    return !symbolicShapeEnabled();
```
Comment (miladm) [inline torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp:159]: * To better understand this condition: the intended behavior is to not fall back to CPU when symbolic shape are enabled. Correct?

* If yes, to above, are you planning to add other dynamic ops like `unique` and `masked_select` to the list? (realizing these ops may be out of scope for this PR)

File: torch/csrc/lazy/core/shape_inference.cpp [hunk 2]
```diff
@@ -244,6 +244,25 @@ std::vector<Shape> compute_shape_min(const at::Tensor & self){
...
+    for (auto dim_size : t.sizes()) {
+      res.emplace_back(Shape(at::kLong, {dim_size}));
+    }
+    return res;
+  }
+  int64_t max_elements = 1;
+  for (auto dim_size : t.sizes()) {
+    max_elements *= dim_size;
+  }
+  return {Shape(at::kLong, {max_elements, (int64_t)t.sizes().size()})};
```
Comment (miladm) [inline torch/csrc/lazy/core/shape_inference.cpp:259]: Please help me understand why we don't do something like this:

`return {Shape(at::kLong, std::vector<int64_t>(t.sizes().begin(), t.sizes().end()))};`

File: torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp (hunk 1 above)
Comment (Gamrix) [reply torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp:159]: That is correct, and yes, we will add at least `unique` to the list, I will need to look at `masked_select` some more.

File: torch/csrc/lazy/core/shape.cpp [hunk 3]
```diff
@@ -108,8 +108,8 @@ void applySymbolicShapesOnLT(
   }
   auto res_symbolic = jit::calculateSymbolicShapesOnOp(&schema, converted_args);
   if (!res_symbolic) {
-    for (int64_t i = 0; i < res_symbolic->size(); i++) {
-      result_shapes[i] = result_shapes[i].with_symbolic_dims(c10::nullopt);
+    for (auto & result_shape : result_shapes) {
+      result_shape = result_shape.with_symbolic_dims(c10::nullopt);
```
Comment (Gamrix) [inline torch/csrc/lazy/core/shape.cpp:112]: Fixed a bug where we crash when there isn't a shape function.

**Comment by Gamrix** [comment]
@Krovatkin Added a test and fixed a bug.

**Review by Krovatkin — APPROVED** [review]
:shipit:

**Pull request closed without merging**